### PR TITLE
[ONNX] Add dim_param support in export with onnx shape inference

### PR DESCRIPTION
**Pull request by BowenBao**
* Support propagating `dim_param` in ONNX by encoding as `ShapeSymbol` in `SymbolicShape` of outputs. If export is called with `dynamic_axes` provided, shape inference will start with these axes set as dynamic.
* Add new test file `test_pytorch_onnx_shape_inference.py`, reusing all test cases from `test_pytorch_onnx_onnxruntime.py`, but focus on validating shape for all nodes in graph. Currently this is not enabled in the CI, since there are still quite some existing issues and corner cases to fix. The test is default to run only at opset 12.
* Bug fixes, such as div, _len, and peephole.cpp passes for PackPadded, and LogSoftmaxCrossEntropy.
* This PR depends on existing PR such as 44332.

### CODE REVIEW

File: scripts/onnx/test.sh [hunk 1]
```diff
@@ -70,4 +70,6 @@ if [[ "$BUILD_ENVIRONMENT" == *ort_test2* ]]; then
     pytest "${args[@]}" \
       "$top_dir/test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset$i"
   done
+  pytest "${args[@]}" \
+    "$top_dir/test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset12_onnx_shape_inference"
```
Comment (spandantiwari) [inline scripts/onnx/test.sh:74]: My suggestion would be create a new file for the shape inference tests. `test_pytorch_onnx_onnxruntime.py` is already loaded with large number of test points, and pulling these into a separate file would be more organized.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -44,6 +46,8 @@ def convert_to_onnx(model, input=None, opset_version=9, example_outputs=None,
                        fixed_batch_size=fixed_batch_size, training=training,
                        onnx_shape_inference=onnx_shape_inference,
                        use_new_jit_passes=use_new_jit_passes)
+    export_time = time.perf_counter() - export_time
+    print('Time to export: ', export_time)
```
Comment (spandantiwari) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:50]: Do we need this in the CI tests?

File: scripts/onnx/test.sh (hunk 1 above)
Comment (spandantiwari) [reply scripts/onnx/test.sh:74]: I see that you already have a new file `test/onnx/test_pytorch_onnx_shape_inference.py`, so my comment above may not be applicable. Still a bit surprised to see this line here.

File: test/onnx/test_pytorch_onnx_shape_inference.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        for out in n.outputs():
+            if not _is_tensor_list(out) and not _is_tensor(out) and not _is_none(out):
+                raise RuntimeError("Output of node is neither type Tensor nor type list of Tensor: ", out)
+            if _is_tensor(out) and out.type().scalarType() is None:
+                raise RuntimeError("Output of node does not have type assigned", out)
+            if _is_tensor(out) and out.type().dim() is None:
+                raise RuntimeError("Output of node does not have shape assigned", out)
+
+
+def run_model_test(self, model, batch_size=2, state_dict=None,
```
Comment (spandantiwari) [inline test/onnx/test_pytorch_onnx_shape_inference.py:32]: Are these `run*` methods quite similar to methods in other test scripts? If they are, would it be possible to refactor them into a common place, e.g. something like `test_helper_utils`.

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 4]
```diff
@@ -310,7 +310,13 @@ void pushPackingPastRnn(Block* b) {
       std::vector<int64_t> new_sizes;
       new_sizes.push_back(*oldType->sizes()[0]);
       new_sizes.push_back(*oldType->sizes()[1]);
-      new_sizes.push_back(rnn->i(attr::hidden_size));
+      if (next->kind() == onnx::Reshape) {
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/peephole.cpp:313]: Interesting...is this fixing an existing bug?

File: torch/csrc/jit/serialization/export.h [hunk 5]
```diff
@@ -21,7 +25,13 @@ namespace jit {
 // file contents being the raw tensor data.
 using RawDataExportMap = std::unordered_map<std::string, at::Tensor>;
 
-TORCH_API std::tuple<std::string, RawDataExportMap> export_onnx(
+using SymbolDimMap = std::map<c10::ShapeSymbol, std::string>;
```
Comment (spandantiwari) [inline torch/csrc/jit/serialization/export.h:28]: I second the choice of using map instead unordered_map to this run-to-run deterministic. This map is small enough where this should not have any significant perf impact.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 6]
```diff
@@ -284,20 +313,19 @@ void SpecialPostProcess(Node* n) {
...
     // get data from value_info and updated original graph.
     for (size_t j = 0; j < clone_node->outputs().size(); ++j) {
       if (clone_node->output(j)->debugName() == v_info.name()) {
-        UpdateTorchValueByOnnxValueInfo(n->output(j), v_info);
+        UpdateTorchValueByOnnxValueInfo(n->output(j), v_info, symbol_map);
       }
     }
   }
-
-  SpecialPostProcess(n);
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:300]: `SpecialPostProcess` not needed anymore?

Comment (spandantiwari) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:300]: Never mind - I see it below.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 7]
```diff
@@ -322,26 +350,103 @@ void ONNXShapeTypeInference(Node* n, int opset_version) {
...
+    std::shared_ptr<Graph>& graph,
+    const std::unordered_map<
+        std::string,
+        std::unordered_map<int64_t, std::string>>& dynamic_axes,
+    const std::vector<std::string>& input_names) {
+  GRAPH_UPDATE("ONNX set dynamic input shape.");
+  auto new_graph = graph->copy();
+  GRAPH_UPDATE("dynamic axes tensor names:", [&]() {
+    std::vector<std::string> res;
+    for (auto it : dynamic_axes) {
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:386]: Suggestion to consider the following STL code for populating `res`:
```
std::vector<std::string> res(dynamic_axes.size());
std::transform(dynamic_axes.begin(), dynamic_axes.end(), res.begin(), [](auto std::pair){return pair.first;});

```

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 8]
```diff
@@ -322,26 +350,103 @@ void ONNXShapeTypeInference(Node* n, int opset_version) {
...
+      res.emplace_back(it.first);
+    }
+    return res;
+  }());
+
+  std::unordered_map<std::string, ::c10::ShapeSymbol> name_to_sym;
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:392]: Should we consider making this a std::map too, just like how you use map above for `SymbolDimMap`?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 9]
```diff
@@ -322,26 +350,103 @@ void ONNXShapeTypeInference(Node* n, int opset_version) {
...
+
+      new_graph->inputs()[i]->setType(
+          input_tensor_type->withSymbolicShapes(::c10::SymbolicShape(shape)));
+    }
+  }
+
+  return new_graph;
+}
+
+std::shared_ptr<Graph> ONNXAssignOutputShape(
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:424]: An aside question: Does the input argument `graph` and the output argument which is a `Graph` pointer have to be `shared_ptr` or can it be `unique_ptr`?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 10]
```diff
@@ -322,26 +350,103 @@ void ONNXShapeTypeInference(Node* n, int opset_version) {
...
+      retval->outputs()[i]->inferTypeFrom(outputs[i]);
+    }
+  }
+  return retval;
+}
+
+std::shared_ptr<Graph> ONNXShapeTypeInference(
+    std::shared_ptr<Graph>& graph,
+    int opset_version) {
+  auto retval = graph->copy();
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:444]: Is this is a deep copy? I am wondering if we need to deep copy every time we update the graph with propagate shape info. I see this above in `ONNXSetDynamicInputShape` and `ONNXAssignOutputShape` as well.

File: torch/csrc/jit/passes/onnx/shape_type_inference.h [hunk 11]
```diff
@@ -8,11 +8,27 @@ namespace jit {
 TORCH_API TypePtr
 MergeInferredType(TypePtr existing_type, TypePtr inferred_type);
 
+TORCH_API std::shared_ptr<Graph> ONNXSetDynamicInputShape(
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/shape_type_inference.h:11]: Some comments to document these methods would be nice here.

File: torch/csrc/jit/python/script_init.cpp [hunk 12]
```diff
@@ -146,11 +146,11 @@ struct PythonResolver : public Resolver {
   ClassTypePtr classType_;
 };
 
-std::shared_ptr<PythonResolver> pythonResolver(ResolutionCallback rcb) {
+std::shared_ptr<PythonResolver> pythonResolver(const ResolutionCallback& rcb) {
```
Comment (spandantiwari) [inline torch/csrc/jit/python/script_init.cpp:149]: These don't seem related. Are these in response to suggestions from clint tool?

File: torch/onnx/symbolic_opset9.py [hunk 13]
```diff
@@ -103,11 +103,11 @@ def mul(g, self, other):
 
 
 def div(g, self, other):
-    return g.op("Div", self, other)
+    return true_divide(g, self, other)
```
Comment (spandantiwari) [inline torch/onnx/symbolic_opset9.py:106]: I am a bit confused as to whether these are needed in this change.

File: torch/csrc/jit/python/script_init.cpp [hunk 14]
```diff
@@ -490,21 +490,6 @@ static std::shared_ptr<Graph> _propagate_and_assign_input_shapes(
   return retval;
 }
 
-static std::shared_ptr<Graph> _assign_output_shapes(
```
Comment (BowenBao) [inline torch/csrc/jit/python/script_init.cpp:493]: Only this part of removing unused code block is related to this PR. The rest of the changes are to pass clang-tidy checks.

File: torch/onnx/symbolic_opset9.py (hunk 13 above)
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py:106]: yes, this is actually from the PR #44831. Included here to pass the scripting test with shape inference enabled for div. I'll mention #44831 as a dependency as well.

File: torch/csrc/jit/python/script_init.cpp (hunk 12 above)
Comment (BowenBao) [reply torch/csrc/jit/python/script_init.cpp:149]: yes, only the below part where i left a comment on is intended change.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 10 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:444]: yea, i think we can update on the original graph, will make the changes.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 9 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:424]: related to the comment below, i think this can change to update graph argument. Regardless, i think unique_ptr may not be applicable, i'm not seeing a single unique_ptr<Graph> occurrence in the code, probably due to its interaction with pybind, that unique_ptr cannot appear as argument.

File: torch/csrc/jit/passes/onnx/peephole.cpp (hunk 4 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/peephole.cpp:313]: yes, the bug was undiscovered before because in unit tests, this output is graph output, so its shape assignment here is overwritten by the output tensor's shape. Now that shape inference is comparing inferred shape with output tensor's shape, this contradiction is discovered.

File: scripts/onnx/test.sh (hunk 1 above)
Comment (BowenBao) [reply scripts/onnx/test.sh:74]: yes, putting it in `test_pytorch_onnx_onnxruntime.py` for one this is a test that runs ort, and then this is temporary and will be removed(or actually, replacing current tests) when shape inference is turned on by default.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:50]: i feel it's a nice to have, makes it easier to monitor how changes affect export time. Perhaps can consider adding something like this to export api in the future.

File: test/onnx/test_pytorch_onnx_shape_inference.py (hunk 3 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_shape_inference.py:32]: These are overloaded(monkey patched) to run specific tests, here running shape verification, and in test_pytorch_onnx_onnxruntime running ORT test.

**Comment by BowenBao** [comment]
rebased to resolve conflict.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 15]
```diff
@@ -82,33 +83,51 @@ TensorTypePtr TorchTensorTypeFromONNX(
...
 
     for (int i = 0; i < onnx_shape.dim_size(); ++i) {
       auto& dim = onnx_shape.dim(i);
       if (dim.has_dim_value()) {
-        sizes.push_back(dim.dim_value());
+        sizes.emplace_back(c10::ShapeSymbol::fromStaticSize(dim.dim_value()));
       } else {
-        // TODO: handle dim_param?
-        return v_type;
+        GRAPH_UPDATE("Got dim_param:", dim.dim_param());
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:94]: Is this needed? Or it's only for debugging?

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 16]
```diff
@@ -320,26 +350,94 @@ void ONNXShapeTypeInference(Node* n, int opset_version) {
...
+        shape[axis] = name_to_sym[name];
+      }
+
+      graph->inputs()[i]->setType(
+          input_tensor_type->withSymbolicShapes(::c10::SymbolicShape(shape)));
+    }
+  }
+}
+
+void ONNXAssignOutputShape(
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:421]: Does this pass support optional outputs represented with Nonetype ?

File: scripts/onnx/test.sh (hunk 1 above)
Comment (spandantiwari) [reply scripts/onnx/test.sh:74]: OK.

File: test/onnx/test_pytorch_onnx_shape_inference.py (hunk 3 above)
Comment (spandantiwari) [reply test/onnx/test_pytorch_onnx_shape_inference.py:32]: OK. Maybe add a comment to this effect here, so that folks are clear that these are "supposed" to be different from similar `run*` functions in other file.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 9 above)
Comment (spandantiwari) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp]: OK.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 15 above)
Comment (spandantiwari) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:94]: My understanding is that we are putting these in for better logging and debugging purpose.

**Review by spandantiwari — APPROVED** [review]
LGTM.

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp (hunk 16 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:421]: no it doesn't. The current code base requires inputs/outputs to be tensors or tuples/lists/dict of tensors.

**Comment by BowenBao** [comment]
@bzinodev This PR has been approved and rebased, please help importing for merge.

**Review by bzinodev — APPROVED** [review]
